JIT 4: jit-aware LQPTranslator and JitOperator

jfrohnhofen opened the pull request:
This PR finally introduces the `JitOperatorWrapper`, that is able to wrap the operators from the previous PR and make them accessible to opossum. This also introduces the `JitAwareLQPTranslator`, that can be used as a drop-in replacement for the existing LQPTranslator and will insert `JitOperatorWrapper` nodes where it sees fit.

mrks commented on src/lib/logical_query_plan/jit_aware_lqp_translator.hpp:
```diff
@@ -0,0 +1,62 @@
+
+/* This class can be used as a drop-in replacement for the LQPTranslator.
+ * The JitAwareLQPTranslator will try to translate multiple AbstractLQPNodes into a single JitOperator, whenever
+ * that is possible and seems beneficial. Otherwise it will fall back to the LQPTranslator.
```
I believe we already started a discussion on this and think we should discuss this in person together with @mrzzzrm. Adding a second translator with unclear responsiblities /will/ lead to inconsistencies one day.

mrks commented on src/lib/logical_query_plan/jit_aware_lqp_translator.hpp:
```diff
@@ -0,0 +1,65 @@
+
+/* This class can be used as a drop-in replacement for the LQPTranslator.
+ * The JitAwareLQPTranslator will try to translate multiple AbstractLQPNodes into a single JitOperatorWrapper, whenever
+ * that is possible and seems beneficial. Otherwise it will fall back to the LQPTranslator.
```
Otherwise**,**

mrks commented on src/lib/logical_query_plan/jit_aware_lqp_translator.hpp:
```diff
@@ -0,0 +1,65 @@
+ *
+ * It works in two steps:
+ * 1) Determine, if we can/should add a JitOperatorWrapper node here and which nodes we can replace:
+ *    Starting from the current node, we perform a BFS through the query tree. For each node we will determine, whether
```
Please write out BFS the first time you use it

mrks commented on src/lib/logical_query_plan/jit_aware_lqp_translator.hpp:
```diff
@@ -0,0 +1,65 @@
+ * It works in two steps:
+ * 1) Determine, if we can/should add a JitOperatorWrapper node here and which nodes we can replace:
+ *    Starting from the current node, we perform a BFS through the query tree. For each node we will determine, whether
+ *    it is jitable (based on the node's type and parameters). We will follow each branch of the tree until we hit a
```
For the whole PR: Please check jitable vs. jittable. My gut likes the second one better.

mrks replied:
http://downloads.bbc.co.uk/skillswise/english/en17suff/factsheet/en17suff-l1-f-suffix-spelling-rules-double-letters.pdf

> 1. For most short (one syllable) words that end in a single consonant ([...]), double the last letter when 
> adding a suffix: e.g. run + ing = running

mrks commented on src/lib/logical_query_plan/jit_aware_lqp_translator.hpp:
```diff
@@ -0,0 +1,65 @@
+ *    it is jitable (based on the node's type and parameters). We will follow each branch of the tree until we hit a
+ *    non-jitable node. Since StoredTableNodes are not jitable, this is guaranteed to happen for all branches.
+ *    All non-jitable nodes encountered this way are stored in a set.
+ *    Once the BFS terminates, we only continue, if the number of jitable nodes is above some threshold and the
```
no comma before if

mrks replied:
What type of threshold?

mrks commented on src/lib/logical_query_plan/jit_aware_lqp_translator.hpp:
```diff
@@ -0,0 +1,65 @@
+ * 1) Determine, if we can/should add a JitOperatorWrapper node here and which nodes we can replace:
+ *    Starting from the current node, we perform a BFS through the query tree. For each node we will determine, whether
+ *    it is jitable (based on the node's type and parameters). We will follow each branch of the tree until we hit a
+ *    non-jitable node. Since StoredTableNodes are not jitable, this is guaranteed to happen for all branches.
```
How do you deal with StoredTableNodes? Why are they not jittable?

jfrohnhofen replied:
They are implicitly jitted, since they provide the data to the operator pipeline. But no dedicated JitOperator instance is created for `StoredTableNodes`.

mrks commented on src/lib/logical_query_plan/jit_aware_lqp_translator.hpp:
```diff
@@ -0,0 +1,65 @@
+ *    We start by adding a JitReadTuple node. This node is passed to all translation functions during the construction
+ *    of further operators. If any jit operator depends on a column or literal value, this value is registered with the
+ *    JitReadTuple operator. The operator returns a JitTupleValue that serves as a placeholder in the requesting
+ *    operator. The JitReadTuple operator will make sure, that the actual value is then accessible through the
```
no comma before that

mrks commented on src/lib/logical_query_plan/jit_aware_lqp_translator.hpp:
```diff
@@ -0,0 +1,65 @@
+ *    JitTupleValue at runtime.
+ *    The output columns are determined by the top-most ProjectionNode. If there is no ProjectionNode, all columns from
+ *    the input node are considered as outputs.
+ *    In case we find any PredicateNode or UnionNode during our traversal, we need to a JitFilter operator.
```
missing word

mrks commented on src/lib/logical_query_plan/jit_aware_lqp_translator.hpp:
```diff
@@ -0,0 +1,65 @@
+  std::shared_ptr<const JitExpression> _translate_to_jit_expression(
+      const std::shared_ptr<PredicateNode>& node, JitReadTuple& jit_source,
+      const std::shared_ptr<AbstractLQPNode>& input_node) const;
+  std::shared_ptr<const JitExpression> _translate_to_jit_expression(
```
Please add empty lines between the methods.

mrks commented on src/lib/logical_query_plan/jit_aware_lqp_translator.hpp:
```diff
@@ -0,0 +1,65 @@
+      const std::shared_ptr<AbstractLQPNode>& node, JitReadTuple& jit_source,
+      const std::shared_ptr<AbstractLQPNode>& input_node) const;
+  std::shared_ptr<const JitExpression> _translate_to_jit_expression(
+      const std::shared_ptr<PredicateNode>& node, JitReadTuple& jit_source,
```
Hmm, this seems dangerous to me. The first two private methods are different only in the first parameter (`AbstractLQPNode` vs. `PredicateNode`). We can have a pointer to an `AbstractLQPNode` that is a `PredicateNode`, but since that is RTTI information, the first method would be called. I think that we should make this more explicit.

mrks commented on src/lib/logical_query_plan/jit_aware_lqp_translator.hpp:
```diff
@@ -0,0 +1,65 @@
+      const AllParameterVariant& value, JitReadTuple& jit_source,
+      const std::shared_ptr<AbstractLQPNode>& input_node) const;
+
+  bool _has_another_condition(const std::shared_ptr<AbstractLQPNode>& node) const;
```
Comments for methods?

mrks commented on src/lib/logical_query_plan/jit_aware_lqp_translator.hpp:
```diff
@@ -0,0 +1,65 @@
+
+namespace opossum {
+
+/* This class can be used as a drop-in replacement for the LQPTranslator.
```
I would rather call it "specialization" instead of "replacement", because it inherits from LQPTranslator

mrks commented on src/lib/operators/jit_operator_wrapper.cpp:
```diff
@@ -0,0 +1,86 @@
+  return std::make_shared<JitOperatorWrapper>(input_left()->recreate(args), _use_jit, _operators);
+}
+
+void JitOperatorWrapper::add_jit_operator(const std::shared_ptr<AbstractJittable>& op) { _operators.push_back(op); }
```
Maybe call the member `_jit_operators`?

mrks commented on src/lib/operators/jit_operator_wrapper.cpp:
```diff
@@ -0,0 +1,86 @@
+
+namespace opossum {
+
+JitOperatorWrapper::JitOperatorWrapper(const std::shared_ptr<const AbstractOperator> left, const bool use_jit,
```
Naming: This is the wrapper for JIT operators - how does `bool use_jit` make sense?

jfrohnhofen replied:
The boolean specifies, whether the pipeline should be compiled or interpreted. How about an enum `ExecutionMode`?

mrks replied:
Sounds good!

mrks commented on src/lib/operators/jit_operator_wrapper.cpp:
```diff
@@ -0,0 +1,86 @@
+std::shared_ptr<AbstractOperator> JitOperatorWrapper::_on_recreate(
+    const std::vector<AllParameterVariant>& args, const std::shared_ptr<AbstractOperator>& recreated_input_left,
+    const std::shared_ptr<AbstractOperator>& recreated_input_right) const {
+  // What should happen here? What are the semantics of recreate?
```
Please clarify your question. What is missing in the documentation of `AbstractOperator::recreate`?

jfrohnhofen replied:
done.

mrks commented on src/lib/operators/jit_operator_wrapper.cpp:
```diff
@@ -0,0 +1,86 @@
+}
+
+// cleanup
+#undef JIT_ABSTRACT_SOURCE_EXECUTE_MANGLED_NAME
```
?

jfrohnhofen replied:
removed

mrks commented on src/lib/operators/jit_operator_wrapper.hpp:
```diff
@@ -0,0 +1,46 @@
+  const std::string description(DescriptionMode description_mode) const final;
+  std::shared_ptr<AbstractOperator> recreate(const std::vector<AllParameterVariant>& args) const final;
+
+  void compile_query();
```
Please document methods

mrks commented on src/lib/sql/sql_pipeline_statement.cpp:
```diff
@@ -16,8 +15,20 @@
+#if HYRISE_JIT_SUPPORT
+using CurrentLQPTranslator = JitAwareLQPTranslator;
+#else
+using CurrentLQPTranslator = LQPTranslator;
```
This should be a runtime setting if JIT support is enabled. Maybe discuss this with @mrzzzrm

jfrohnhofen replied:
I never liked that compile-time mechanism either, but it kind of worked for me so far. Where would that setting come from and how would it get propagated to the pipeline statement? @mrzzzrm Any ideas?

mrzzzrm replied:
You could make `_use_jit` a parameter of the SQLPipeline/SQLPipelineBuilder and then pick the correct Translator at runtime, while Assert()ing, that `_use_jit` is only set to true when `HYRISE_JIT_SUPPORT`.

jfrohnhofen replied:
I added an `lqp_translator` parameter to SQLPipeline/SQLPipelineBuilder analog to the optimizer and also added a check of `HYRISE_JIT_SUPPORT ` to the constructor of the jit-aware translator.

mrks commented on src/lib/logical_query_plan/jit_aware_lqp_translator.cpp:
```diff
@@ -0,0 +1,273 @@
+#include "jit_aware_lqp_translator.hpp"
```
What are your plans for testing this?

mrks commented on src/lib/logical_query_plan/jit_aware_lqp_translator.cpp:
```diff
@@ -0,0 +1,273 @@
+  std::unordered_set<std::shared_ptr<AbstractLQPNode>> input_nodes;
+
+  // Traverse query tree until a non-jitable nodes is found in each branch
+  _breadth_first_search(node, [&](auto& current_node) {
```
Is the BFS aspect relevant for the result or is it an irrelevant implementation detail?

mrks replied:
(moving this to the discussion in the `_breadth_first_search` implementation)

mrks commented on src/lib/logical_query_plan/jit_aware_lqp_translator.cpp:
```diff
@@ -0,0 +1,273 @@
+
+  // It does not make sense to create a JitOperatorWrapper for fewer than 2 LQP nodes,
+  // but we may want a better heuristic here
+  if (num_jittable_nodes < 2 || input_nodes.size() != 1) {
```
This looks like you cannot deal with `OR`?

jfrohnhofen replied:
You mean the `input_nodes.size() != 1` part? `OR`s are represented by `UnionNode`s, but both inputs to the union eventually refer to the same node (i.e., a diamond pattern).

mrks commented on src/lib/logical_query_plan/jit_aware_lqp_translator.cpp:
```diff
@@ -0,0 +1,273 @@
+
+  const auto input_node = *input_nodes.begin();
+
+  auto jit_operator = std::make_shared<JitOperatorWrapper>(translate_node(input_node));
```
This could use a comment that you are not wrapping the input_node but are only using it as the input to the wrapper.

mrks commented on src/lib/logical_query_plan/jit_aware_lqp_translator.cpp:
```diff
@@ -0,0 +1,273 @@
+    if (filter_node != input_node) {
+      // However, if we do need to filter, we first convert the filter node to a JitExpression ...
+      const auto expression = _translate_to_jit_expression(filter_node, *read_tuple, input_node);
+      // make sure that expression get computed ...
```
**the** expression get**s** computed

mrks commented on src/lib/logical_query_plan/jit_aware_lqp_translator.cpp:
```diff
@@ -0,0 +1,273 @@
+    }
+
+    // Identify the top-most projection node to determine the output columns
+    auto top_projection = node;
```
Naming: This might not be a projection node

mrks commented on src/lib/logical_query_plan/jit_aware_lqp_translator.cpp:
```diff
@@ -0,0 +1,273 @@
+      top_projection = top_projection->left_input();
+    }
+
+    // Add a compute operator for each output column that computes the column value.
```
"that computes the column value" is a bit unclear for me. Do you mean "that is calculated during execution time (i.e., that is not a column from a stored table)"?

mrks commented on src/lib/logical_query_plan/jit_aware_lqp_translator.cpp:
```diff
@@ -0,0 +1,273 @@
+    }
+
+    // Add a compute operator for each output column that computes the column value.
+    auto write_table = std::make_shared<JitWriteTuple>();
```
Naming mismatch: write table vs. write tuple? I think it should be JitWriteTable or JitWriteTuples

mrks replied:
ping?

mrks commented on src/lib/logical_query_plan/jit_aware_lqp_translator.cpp:
```diff
@@ -0,0 +1,273 @@
+    }
+    jit_operator->add_jit_operator(write_table);
+    return jit_operator;
+  } catch (const std::exception& exception) {
```
When does this happen? I am not a fan of using exceptions for expected control flow. Can we do this in a way where we check the return values of the methods or would that make things more complicated?

mrzzzrm replied:
I'd second Markus on his

jfrohnhofen replied:
This can happen in a number of cases. E.g. if some expression type or operand is not supported, if an unexpected node type is encountered or if the source of a column cannot be located.
We could add additional control flow for all of this, but the checks are quite spread across the code, so I am not sure that would simplify things.

mrzzzrm replied:
Without having looked into it deeply, I think the preferred way should be to make explicit checks for what nodes/expressions are supported and invoke the normal LQPTranslator in a controlled way rather than doing this via an exception. On second thought I am pretty convinced the exception based way is very, very unclean.
std::exception could be anything, e.g., bad_alloc, some important internal Assertion. Just falling back to LQPTranslator would mask away important error messages then.

jfrohnhofen replied:
I agree that exceptions are unclean. IMO there are two ways to do explicit checks:
1) do all checks first, decide which translator to use and then do the translation in a subsequent step. This leads to duplicated code, since node types and parameters have to be checked / dispatched twice.
2) do checks and translation in a single step. However, this requires a way to cancel query translation in the middle and exceptions seem to be the cleanest way to achieve that.
I don't really like either option, but cannot think of a good third option. Do you have any ideas?
We could of course switch to a dedicated opossum::translation_exception if that is your primary concern.

mrzzzrm replied:
I think 1. should be the way to go. If some unavoidable code duplication comes out of it that is acceptable, imo. 
If we allow `opossum::translation_exception` in this case we'll have countless `opossum::do_my_control_flow_for_me_exception` within a years time as people adopt this and I think we don't want to get into that position.

mrks replied:
How about checking the return value for `nullptr`? If something can't be translated, I get a `nullptr` and that way I know to use the non-JIT path. That should keep us from duplicating any code.

mrks replied:
I don't even think that that would be hacky. It's more of a "please translate this" - "no".

mrzzzrm replied:
Just an example why I don't like exceptions for control flow: You can't debug your program with breakpoint-on-exception because then it breaks in countless places where no error ever occured and you never actually get to step to the actual exception you're debugging because you give up before that. Sure, you could do breakpoint-on-specific-exception-type, but its still a hazard.

mrks commented on src/lib/logical_query_plan/jit_aware_lqp_translator.cpp:
```diff
@@ -0,0 +1,273 @@
+      // If there are further conditions (either PredicateNodes of UnionNodes) down the line, we need a logical AND here
+      if (_has_another_condition(node->left_input())) {
+        left = _translate_to_jit_expression(std::dynamic_pointer_cast<PredicateNode>(node), jit_source, input_node);
+        right = _translate_to_jit_expression(node->left_input(), jit_source, input_node);
```
I am confused of your use of left and right here. What exactly happens in this method?

jfrohnhofen replied:
Added comments to clarify.

mrks commented on src/lib/logical_query_plan/jit_aware_lqp_translator.cpp:
```diff
@@ -0,0 +1,273 @@
+    case ExpressionType::GreaterThanEquals:
+    case ExpressionType::Like:
+    case ExpressionType::NotLike:
+      right = _translate_to_jit_expression(node->value(), jit_source, input_node);
```
Is `right` the right hand side of the comparison? Usually, when we talk about LQPs, left and right are tables. Maybe that's where my confusion comes from.

jfrohnhofen replied:
Yes. What alternative would you prefer? `left_side` / `left_input` / `left_child` / `lhs`?

mrks commented on src/lib/logical_query_plan/jit_aware_lqp_translator.cpp:
```diff
@@ -0,0 +1,273 @@
+  }
+}
+
+bool JitAwareLQPTranslator::_has_another_condition(const std::shared_ptr<AbstractLQPNode>& node) const {
```
The name of this method is not really speaking. Not all nodes have conditions. And what does "another" mean? Another than... what?

jfrohnhofen replied:
Renamed the method and added some more comments to explain. Please have another look.

mrks commented on src/lib/logical_query_plan/jit_aware_lqp_translator.cpp:
```diff
@@ -0,0 +1,273 @@
+  }
+}
+
+void JitAwareLQPTranslator::_breadth_first_search(
```
A search method that doesn't give you a result? If `func` is expected to keep track of the results, it is not a search, maybe an `apply`.

jfrohnhofen replied:
IMO `apply` does not really capture that multiple elements might be affected. I went with `visit` for now. Ok?

mrzzzrm commented on src/lib/logical_query_plan/jit_aware_lqp_translator.cpp:
```diff
@@ -0,0 +1,273 @@
+  auto read_tuple = std::make_shared<JitReadTuple>();
+  jit_operator->add_jit_operator(read_tuple);
+
+  auto filter_node = node;
```
maybe just filter_node = node->left_input instead of that weird first loop condition?

jfrohnhofen replied:
This would not work if there are (for some reason) multiple `ProjectionNode`s

mrzzzrm commented on src/lib/logical_query_plan/jit_aware_lqp_translator.cpp:
```diff
@@ -0,0 +1,273 @@
+
+    // Identify the top-most projection node to determine the output columns
+    auto top_projection = node;
+    while (top_projection != input_node && top_projection->type() != LQPNodeType::Projection) {
```
What happens if you don't find a Projection? LQPs are perfectly valid without them so you cannot expect to always find one.

jfrohnhofen replied:
I have never seen such an LQP. Is this a theoretical/hypothetical possibility or does this really happen? All I need for my translation is a way to get all columns that should be part of the output table. If there is no ProjectionNode the correct behavior IMO would be to use all inout columns. Is there a generic interface to get all columns of an LQP node? I can only see a way to get `LQPColumnReferences` but those dont allow access to data types etc.

mrzzzrm replied:
First, you can't impose any "impossible" LQP layout. Everything is possible and even Projections might be removed by some rule (e.g. SELECT * FROM ... WHERE ...; doesn't require a Projection.

Not sure what you mean by "get all columns"... The LQPColumnReferences are just that. They either point to a StoredTableNode, from whose table you can figure out the data type, or to a expression in a AggregateNode/ProjectionNode and you would somehow need to figure out the result data type of that expression.

mrzzzrm replied:
(It's currently hazardous to get the data type of any arbitrary column in the LQP, I would agree to that. I'm refactoring that right now, but those changes won't come in anytime soon)

mrzzzrm commented on src/lib/logical_query_plan/jit_aware_lqp_translator.cpp:
```diff
@@ -0,0 +1,273 @@
+      return std::make_shared<JitExpression>(left, ExpressionType::Or, right, jit_source.add_temporary_value());
+
+    case LQPNodeType::Projection:
+      // We don't care about projection nodes here, since they do not perform any tuple filtering
```
Is it really safe to discard Projections? They could compute values used higher up in the LQP, after all.

jfrohnhofen replied:
It is safe in this place, since we are only interested in tuple filtering here.
If we work on a computed column value, this will be handled when translating the corresponding `LQPColumnReference`.

mrzzzrm commented on src/lib/logical_query_plan/jit_aware_lqp_translator.cpp:
```diff
@@ -0,0 +1,273 @@
+    const auto lqp_expression = projection_node->column_expressions()[lqp_column_reference.original_column_id()];
+    return _translate_to_jit_expression(*lqp_expression, jit_source, input_node);
+  } else {
+    Fail("Could not get column from anywhere.");
```
That means you can't deal with columns created by an Aggregate expression. Will this Fail get reached if a column from an AggregateNode is referenced?

mrzzzrm commented:
There are three LQP Node Types "producing" columns: StoredTableNode, ProjectionNode and AggregateNode. Everyone else just forwards already existing Columns.

mrks commented on src/lib/logical_query_plan/jit_aware_lqp_translator.cpp:
```diff
@@ -0,0 +1,294 @@
+  }
+
+  // Add a compute operator for each computed output column (i.e., a column that is not from a stored table).
+  auto write_table = std::make_shared<JitWriteTuple>();
```
reopening: "Naming mismatch: write table vs. write tuple? I think it should be JitWriteTable or JitWriteTuples"

mrks commented on src/lib/logical_query_plan/jit_aware_lqp_translator.cpp:
```diff
@@ -0,0 +1,294 @@
+  return jit_operator;
+}
+
+std::shared_ptr<const JitExpression> JitAwareLQPTranslator::_translate_node_to_jit_expression(
```
This can also return a `nullptr`. Why is the one before "try" and this one not?

mrks commented on src/lib/logical_query_plan/jit_aware_lqp_translator.cpp:
```diff
@@ -0,0 +1,294 @@
+      // and the input to the predicate node becomes the right side of the conjunction.
+      // If the PredicateNode has no further conditions (i.e., all tuples pass into the PredicateNode unconditionally),
+      // there is no need to create a dedicated AND-expression.
+      if (_subquery_filters_tuples(node->left_input())) {
```
Did we have that in the version that I reviewed as well? I think it is confusing as we already have "subqueries" in SQL, and they are something different. Maybe `_input_is_filtered`?

mrks commented on src/lib/logical_query_plan/jit_aware_lqp_translator.cpp:
```diff
@@ -0,0 +1,294 @@
+  }
+}
+
+std::shared_ptr<const JitExpression> JitAwareLQPTranslator::_translate_predicate_to_jit_expression(
```
Same "try" discussion.

mrks commented on src/lib/logical_query_plan/jit_aware_lqp_translator.cpp:
```diff
@@ -0,0 +1,294 @@
+  }
+}
+
+std::shared_ptr<const JitExpression> JitAwareLQPTranslator::_translate_expression_to_jit_expression(
```
Could you check all methods if they should be "try" or not?

mrks commented on src/lib/logical_query_plan/jit_aware_lqp_translator.hpp:
```diff
@@ -0,0 +1,87 @@
+ *
+ * It works in two steps:
+ * 1) Determine, if we can/should add a JitOperatorWrapper node here and which nodes we can replace:
+ *    Starting from the current node, we perform a breadth-first search through the query tree. For each node we will determine, whether
```
stray commas before "if" and "whether"

mrks commented:
@mrzzzrm - what do you think?

mrks commented:
@FabianWiebe - did you have a chance to look at this?

mrks commented on src/lib/logical_query_plan/jit_aware_lqp_translator.hpp:
```diff
@@ -0,0 +1,88 @@
+ *    by a helper method first. We then add a JitCompute operator to our chain and use its result value instead of the
+ *    original non-primitive value.
+ */
+class JitAwareLQPTranslator final : public LQPTranslator {
```
Sounds reasonable.

mrks commented:
@mrzzzrm - I absolutely agree that we need tests for this. As far as I remember, those already exist, but are not in the code yet because they rely on future PRs. This is not the best way of doing things and hopefully won't happen again. Still, I think we should get this in so that we can work on the final JIT PRs while Johannes can still respond.

mrzzzrm commented:
Alright, have my approval -  we can talk about the directory structure separately.

jfrohnhofen commented:
I added tests for previous Hyrise code as a separate PR #864.
I will add tests for this PR (hopefully) in the next 2-3 days and will continue working on PR5 after that.
I don't mind leaving the PR open until then, but fell free to merge this and I will open a separate PR for the tests.

@mrzzzrm regarding directory structure: I moved to the `JitAwareLQPTranslator` to the `src/lib/operators/jit_operator/..` directory, so that all JIT-related code now lives in this directory.
But I think you suggested moving all JIT code to `src/lib/jit/..`. While this bundles all JIT code nicely, this spreads operator code across two locations. I will move the code accordingly as soon as we are all on the same page. I am happy to do it either way, what is your opinion @mrks?

And finally: The `JitAwareLQPTranslator` can be activated in the `SQLPipelineBuilder` but is deactivated by default (i.e., there is no mechanism/flag/switch in any binary that activates the JIT). Should this be added and if so where/how?
I would suggest adding a command to the console.
We should also run the `CompareToSQLite` tests with activated JIT.
What would be the easiest way to achieve this? Duplicate the entire test? Is there an easy way to run the test (or all tests) with two configurations?

mrzzzrm commented:
> But I think you suggested moving all JIT code to src/lib/jit/... While this bundles all JIT code nicely, this spreads operator code across two locations. I will move the code accordingly as soon as we are all on the same page. I am happy to do it either way, what is your opinion @mrks?

I'd prefer /src/lib/jit, as I feel the "jit" attribute pulls things together tighter than the "operator" attribute. And it might confuse newcomers less. But I have no strong feelings about this.

> We should also run the CompareToSQLite tests with activated JIT. What would be the easiest way to achieve this? Duplicate the entire test? Is there an easy way to run the test (or all tests) with two configurations?

Tagging @lanice but that's something we should handle in another issue, probably?
